### PR TITLE
feat: Allow composite commands to be overridden

### PR DESCRIPTION
**Pull request by johnmcollier**
**What type of PR is this?**
/kind feature
/area devfile

**What does does this PR do / why we need it**:
Adds support for overriding composite commands in parent devfiles. Also fixes a bug where parent commands could not be overridden if the parent command contained upper case letters.

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/3759

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
Sample devfile:
```
schemaVersion: 2.0.0
metadata:
  name: nodejs
  version: 1.0.0
parent:
  uri: https://raw.githubusercontent.com/openshift/odo/master/tests/examples/source/devfiles/nodejs/devfileCompositeCommands.yaml
  commands:
  - id: buildAndMkdir
    composite:
         label: Build and Mkdir
         commands:
           - createfile
           - install
         parallel: false
         group: 
            kind: build
            isDefault: true
commands:
  - id: createfile
    exec:
      component: runtime
      commandLine: touch /projects/testfile
      workingDir: ${PROJECTS_ROOT}
      group:
        kind: build
        isDefault: false
```

Output should have:
```
Executing devfile commands for component nodejs
 ✓  Executing createfile command "touch /projects/testfile" [300ms]
 ✓  Executing install command "npm install" [2s]
 ✓  Executing run command "npm start" [1s]
```

### CODE REVIEW

File: pkg/devfile/parser/types.go [hunk 1]
```diff
@@ -65,22 +65,46 @@ func (d DevfileObj) OverrideComponents(overridePatch []common.DevfileComponent)
...
-				merged, err := handleMerge(originalCommand.Exec, patchCommand.Exec, common.Exec{})
-				if err != nil {
-					return err
-				}
-
-				err = json.Unmarshal(merged, &updatedExec)
-				if err != nil {
-					return errors.Wrap(err, "failed to unmarshal override commands")
+				var devfileCommand common.DevfileCommand
+
```
Comment (adisky) [inline pkg/devfile/parser/types.go:75]: can we break the composite command override into a util function?

Comment (johnmcollier) [reply pkg/devfile/parser/types.go]: Updated, thanks @adisky. I've kept the unit tests that I added as part of TestDevfileObj_OverrideCommands, however.

File: pkg/devfile/parser/types.go [hunk 2]
```diff
@@ -90,6 +105,23 @@ func (d DevfileObj) OverrideCommands(overridePatch []common.DevfileCommand) erro
 	return nil
 }
 
+// overrideCompositeCommand overrides the given parent compoosite commmand
```
Comment (mik-dass) [inline pkg/devfile/parser/types.go:108]: composite

File: pkg/devfile/parser/types.go [hunk 3]
```diff
@@ -62,25 +62,40 @@ func (d DevfileObj) OverrideComponents(overridePatch []common.DevfileComponent)
...
+						return errors.Wrap(err, "failed to unmarshal override commands")
+					}
+					devfileCommand = common.DevfileCommand{Id: patchCommand.Id, Exec: &updatedExec}
+				} else if patchCommand.Composite != nil && originalCommand.Composite != nil {
+					devfileCommand, err = overrideCompositeCommand(patchCommand, originalCommand)
+					if err != nil {
+						return err
+					}
+				} else {
+					// If the original command and patch command are different types, then we can't patch, so throow an error
```
Comment (mik-dass) [inline pkg/devfile/parser/types.go:94]: throw

**Comment by johnmcollier** [comment]
@mik-dass Thanks for reviewing, fixed the typos! The "o" key on my laptop has been acting up lately 😅

**Review by mik-dass — APPROVED** [review]
/approve

**Review by cdrage — APPROVED** [review]
Code LGTM!

/lgtm

**Review by maysunfaisal — APPROVED** [review]

**Comment by kadel** [comment]
re-adding 
/lgtm
after rebase

**Pull request merged**